### PR TITLE
Add shared monorepo build setup

### PR DESCRIPTION
**Pull request by huntie**
Summary:
## Context

RFC: Decoupling Flipper from React Native core: https://github.com/react-native-community/discussions-and-proposals/pull/641

## Changes

To support incoming new React Native packages around debugging (including migrating over [`@react-native-community/cli-plugin-metro`](https://github.com/react-native-community/cli/tree/main/packages/cli-plugin-metro)) — which target Node.js and require a build step, this PR adds a minimal shared build setup across the `react-native` monorepo.

The setup is closely inspired/based on the build scripts in Jest, Metro, and React Native CLI — and is a simple set of script wrappers around Babel and Chokidar. These are available as build commands at the root of the repo:

- `yarn build`
- `yarn clean`
- `yarn watch`

image

Build configuration (i.e. Babel config) is shared as a set standard across the monorepo, and **packages are opted-in to requiring a build**, configured in `scripts/build.config.js`.

```js
const buildConfig /*: BuildConfig */ = {
  // The packages to include for build and their build options
  packages: {
    'dev-middleware': {target: 'node'},
  },
};
```

For now, there is a single `target: 'node'` option — this is necessary as `react-native`, unlike the above other projects, is a repository with packages targeting several runtimes. We may, in future, introduce a build step for other, non-Node, packages — which may be useful for things such as auto-generated TypeScript definitions.

## Current integration points

- `.circleci/config.yml` — Added to `run_yarn` (with project install).
- Root `yarn start` — Convenience for local developers.
- `yarn start` within `packages/rn-tester` — Convenience for local developers.

**Planned**: I will be adding notes on `yarn build` to the following "Contributing" website pages:

- https://reactnative.dev/contributing/overview#contributing-code
- https://reactnative.dev/contributing/how-to-open-a-pull-request#chapter-ii-implementing-your-changes

**This is not load bearing quite yet**: There are not yet any built packages added to the dependencies of `packages/react-native/`, so this will be further tested in a later PR (internally stacked as the next change).

### Alternative designs

**Per-package config file**

Replace `scripts/build/config.js` with a package-defined key in in `package.json`, similar to Jest's [`publishConfig`](https://github.com/jestjs/jest/blob/1f019afdcdfc54a6664908bb45f343db4e3d0848/packages/jest-cli/package.json#L87C3-L89C4).

```json
"buildConfig": {
  "type": "node"
},
```

This would be the only customisation required, with a single Babel config still standardised. Another option this might receive in future is `enableTypeScriptCodgeen`.

**Rollup**

More sophisticated build tool for Node.js, used by the React codebase (albeit within a custom script setup as well).

**Lerna and Nx**

- Most sophisticated setup enabling caching and optimised cloud runs.
- Probably the most likely thing we'll move towards at a later stage.

Changelog: [Internal]

Differential Revision: D46759508

### CODE REVIEW

**Comment by huntie** [comment]
Note: Includes https://github.com/facebook/react-native/pull/38194, will be rebased.

File: packages/dev-middleware/package.json [hunk 1]
```diff
@@ -18,25 +18,12 @@
   "files": [
     "dist"
   ],
-  "scripts": {
-    "build": "yarn clean && babel src --out-dir dist",
-    "dev": "babel src --out-dir dist --source-maps --watch",
-    "clean": "rimraf dist",
-    "prepare": "yarn build"
-  },
```
Comment (Pranav-yadav) [inline packages/dev-middleware/package.json:26]: are we going to publish this package? If yes, how do we want to provide these scripts and devDeps. to consumer. Context: There's no `"private": true` field...

Comment (huntie) [reply packages/dev-middleware/package.json:26]: Yes this package will be published (and depended on in upcoming changes). These `scripts` are now managed at the repo level.

Comment (Pranav-yadav) [reply packages/dev-middleware/package.json:26]: Good to know. 👍

File: package.json [hunk 2]
```diff
@@ -11,8 +11,12 @@
     "outputName": "js-test-results.xml"
   },
   "scripts": {
-    "start": "cd packages/rn-tester && npm run start",
+    "postinstall": "yarn build",
```
Comment (Pranav-yadav) [inline package.json:14]: will u pls explain how this is going to help ?

File: package.json [hunk 3]
```diff
@@ -11,8 +11,12 @@
...
+    "start": "yarn build && cd packages/rn-tester && npm run start",
     "android": "cd packages/rn-tester && npm run android",
+    "build": "node ./scripts/build/build.js",
+    "clean": "node ./scripts/build/clean.js",
+    "watch": "yarn build && node ./scripts/build/watch.js",
```
Comment (Pranav-yadav) [inline package.json:19]: I think there's no need for an extra `watch` script.
We can just have a "single build script" which takes `--watch`.
That would be awesome.
WDYT?

File: scripts/build/build.js [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+const PACKAGES_DIR /*: string */ = path.resolve(__dirname, '../../packages');
+const SRC_DIR = 'src';
+const BUILD_DIR = 'dist';
+const JS_FILES_PATTERN = '**/*.js';
+const IGNORE_PATTERN = '**/__{tests,mocks,fixtures}__/**';
+
+const config = {
+  allowPositionals: true,
+  options: {
+    help: {type: 'boolean'},
```
Comment (Pranav-yadav) [inline scripts/build/build.js:31]: here , adding `watch` option, could make things better, as I said above.

File: scripts/build/clean.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  Usage: node ./scripts/build/clean.js
+
+  Clean build directories for all packages defined in ./scripts/build/config.js.
+    `);
+    process.exitCode = 0;
+    return;
+  }
+
+  for (const packageName of Object.keys(buildConfig.packages)) {
+    fs.rmSync(path.join(PACKAGES_DIR, packageName, BUILD_DIR), {
```
Comment (Pranav-yadav) [inline scripts/build/clean.js:40]: can this fail or block, if the stuff it's trying to rm is in use?

**Pull request closed without merging**